File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/core/DefaultManifest.java
package com.ririnto.sinon.harness.core;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import tools.jackson.databind.JsonNode;
import java.util.List;

/**
 * Default manifest facade backed by a Jackson JSON node.
 */
public final class DefaultManifest implements Manifest {
    private final JsonNode raw;

    /**
     * Creates a manifest facade.
     *
     * @param raw raw manifest JSON node
     */
    public DefaultManifest(JsonNode raw) {
        this.raw = raw;
    }

    @Override
    public JsonNode raw() {
        return raw;
    }

    @Override
    public boolean isEnabled(String category) {
        return HarnessCheckHelper.applies(raw, category);
    }

    @Override
    public Severity severityOf(String category) {
        return Severity.valueOf(HarnessCheckHelper.getSeverity(raw, category));
    }

    @Override
    public List<String> stringArray(String category, String key) {
        final JsonNode section = categoryObject(category);
        if (section == null) {
            return List.of();
        }
        final JsonNode parameters = section.get("parameters");
        if (parameters == null) {
            return List.of();
        }
        return HarnessCheckHelper.extractPaths(parameters.get(key));
    }

    @Override
    public String stringValue(String category, String key) {
        final JsonNode section = categoryObject(category);
        if (section == null) {
            return "";
        }
        final JsonNode parameters = section.get("parameters");
        if (parameters == null) {
            return "";
        }
        final JsonNode value = parameters.get(key);
        return value != null && value.isTextual() ? value.asText() : "";
    }

    @Override
    public JsonNode categoryObject(String category) {
        final JsonNode section = raw.get(category);
        return section != null && section.isObject() ? section : null;
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/core/DefaultManifest.kt
package com.ririnto.sinon.harness.core

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Default implementation of the Manifest interface.
 */
class DefaultManifest(override val raw: JsonObject) : Manifest {
    override fun isEnabled(category: String): Boolean =
        categoryField(category, "enabled")
            ?.jsonPrimitive
            ?.contentOrNull
            ?.toBooleanStrictOrNull()
            ?: false

    override fun severityOf(category: String): Severity {
        val severityName =
            categoryField(category, "severity")
                ?.jsonPrimitive
                ?.contentOrNull
                ?: return Severity.ERROR
        return Severity.entries.firstOrNull { entry -> entry.name.equals(severityName, ignoreCase = true) } ?: Severity.ERROR
    }

    override fun stringArray(category: String, key: String): List<String> =
        categoryField(category, key)
            ?.jsonArray
            ?.mapNotNull { elem -> elem.jsonPrimitive.contentOrNull }
            ?: emptyList()

    override fun stringValue(category: String, key: String): String =
        categoryField(category, key)
            ?.jsonPrimitive
            ?.contentOrNull
            ?: ""

    override fun categoryObject(category: String): JsonObject? =
        raw[category]?.jsonObject

    /**
     * Retrieves a field from a category's JSON object.
     *
     * @param category The category name to access.
     * @param key The field key within the category.
     * @return The field as a JsonElement, or null if the category or field does not exist.
     */
    private fun categoryField(category: String, key: String): JsonElement? =
        raw[category]?.jsonObject?.get(key)
}
